### DISCUSSION TITLE
Can GLICOL run on an embedded Arm Cortex M4/M7?

### DISCUSSION THREAD

**Discussion by znmeb** [chaosprint/glicol#77] [Q&A] [answered]
I've got a couple of Teensy 4.1 boards and a couple of Electro-Smith Daisy units. The Daisy ones have audio hardware; the Teensys are just the bare boards, although I could buy audio add-ons for them if they'll run GLICOL. Both have SD card capability.

The Teensy is a 600 MHz Arm Cortex M7 with 7936K flash and 1024K RAM. The Daisy is a 480 MHz Arm Cortex M7. I'm not sure how much flash and RAM the Daisy has; it's not on their website. Both are programmable / flashable via the Arduino IDE. Can GLICOL run on either or both of these?

**Comment by chaosprint** [comment, marked as answer]
Thank you for your question.
I also have STM32F3 and some ESP boards.
In theory it should be able to, but it may require some hacking for now. I will try to test and make the process easier in the coming days.
The key is to make Glicol support `no_std`. Yet two dependencies of Glicol, `petgraph` and `pest` have not supported `no_std` in main releases. There are already some successful forks of these two packages that successfully implemented `no_std` but not yet merged.
https://github.com/pest-parser/pest/pull/498
https://github.com/petgraph/petgraph/pull/370
I will test by depending on the forked git soon.

You get already get started by reading:
https://github.com/rust-embedded/awesome-embedded-rust
Especially the:
[The Embedded Rust Book](https://rust-embedded.github.io/book/)
For audio, I am pretty sure that DASP can support `no_std`:
https://github.com/RustAudio/dasp
For Bela or Pi, it should be more straightforward, although I will still need to change things to support features like ADC on Glicol.
Yet to use Glicol syntax or Glicol synth on `no_std` env, please stay tuned.

**Reply by znmeb** [reply]
Thanks! I have three Belas - two of the full-sized ones and a mini with the eight-channel add-on. I'll start with a Bela to get the toolchain set up. I also have a Jetson AGX Xavier I can use for cross-compiling if necessary - full Ubuntu 18.04 and soon 20.04.

**Reply by chaosprint** [reply]
> Thanks! I have three Belas - two of the full-sized ones and a mini with the eight-channel add-on. I'll start with a Bela to get the toolchain set up. I also have a Jetson AGX Xavier I can use for cross-compiling if necessary - full Ubuntu 18.04 and soon 20.04.

You can start with these repos:
https://github.com/andrewcsmith/bela-rs
https://github.com/padenot/bela-sys
https://github.com/andrewcsmith/bela-sys
And try to get a hello world tone there.

**Reply by znmeb** [reply]
Thanks!

**Reply by chaosprint** [reply]
I got some updates here:
https://github.com/chaosprint/glicol/tree/main/rs/bela

@znmeb

**Comment by znmeb** [comment]
Thanks! I guess I need to buy one of those maker kits with potentiometers. :-) I already have three Belas and plenty of breadboards.

**Comment by strickinato** [comment]
It looks `no_std` support on the related dependencies have been merged 😄 

https://github.com/pest-parser/pest/pull/498
https://github.com/petgraph/petgraph/pull/370

Is there a trodden path for getting glicol running on a teensy?

**Reply by chaosprint** [reply]
I am now rewriting the parser with chumsky https://github.com/zesterer/chumsky
embedded, cli, wasm will all be first class citizens.
So it won't be long to see a working demo

**Reply by strickinato** [reply]
🙌-- I'll keep an eye out.

**Reply by znmeb** [reply]
Cool!!
